cartographer_grpc/local_trajectory_uploader.cc

```diff
@@ -69,57 +69,88 @@ void LocalTrajectoryUploader::ProcessSendQueue() {
                          data_message.get())) {
         ProcessOdometryDataMessage(odometry_data);
+      } else if (auto *local_slam_result_data =
+                     dynamic_cast<proto::AddLocalSlamResultDataRequest *>(
+                         data_message.get())) {
+        ProcessLocalSlamResultDataMessage(local_slam_result_data);
       } else {
         LOG(FATAL) << "Unknown message type: " << data_message->GetTypeName();
       }
     }
   }
 }
 
+void LocalTrajectoryUploader::ProcessSensorMetadata(
+    proto::SensorMetadata *sensor_metadata) {
+  int cloud_trajectory_id =
+      local_to_cloud_trajectory_id_map_.at(sensor_metadata->trajectory_id());
+  sensor_metadata->set_trajectory_id(cloud_trajectory_id);
+}
+
 void LocalTrajectoryUploader::ProcessFixedFramePoseDataMessage(
-    const proto::AddFixedFramePoseDataRequest *data_request) {
+    proto::AddFixedFramePoseDataRequest *data_request) {
   if (!fixed_frame_pose_writer_.client_writer) {
     fixed_frame_pose_writer_.client_writer =
         service_stub_->AddFixedFramePoseData(
             &fixed_frame_pose_writer_.client_context,
             &fixed_frame_pose_writer_.response);
     CHECK(fixed_frame_pose_writer_.client_writer);
   }
+  ProcessSensorMetadata(data_request->mutable_sensor_metadata());
   fixed_frame_pose_writer_.client_writer->Write(*data_request);
 }
 
 void LocalTrajectoryUploader::ProcessImuDataMessage(
-    const proto::AddImuDataRequest *data_request) {
+    proto::AddImuDataRequest *data_request) {
   if (!imu_writer_.client_writer) {
     imu_writer_.client_writer = service_stub_->AddImuData(
         &imu_writer_.client_context, &imu_writer_.response);
     CHECK(imu_writer_.client_writer);
   }
+  ProcessSensorMetadata(data_request->mutable_sensor_metadata());
   imu_writer_.client_writer->Write(*data_request);
 }
 
 void LocalTrajectoryUploader::ProcessOdometryDataMessage(
-    const proto::AddOdometryDataRequest *data_request) {
+    proto::AddOdometryDataRequest *data_request) {
   if (!odometry_writer_.client_writer) {
     odometry_writer_.client_writer = service_stub_->AddOdometryData(
         &odometry_writer_.client_context, &odometry_writer_.response);
     CHECK(odometry_writer_.client_writer);
   }
+  ProcessSensorMetadata(data_request->mutable_sensor_metadata());
   odometry_writer_.client_writer->Write(*data_request);
 }
 
+void LocalTrajectoryUploader::ProcessLocalSlamResultDataMessage(
+    proto::AddLocalSlamResultDataRequest *data_request) {
+  if (!local_slam_result_writer_.client_writer) {
+    local_slam_result_writer_.client_writer =
+        service_stub_->AddLocalSlamResultData(
+            &local_slam_result_writer_.client_context,
+            &local_slam_result_writer_.response);
+    CHECK(local_slam_result_writer_.client_writer);
+  }
+  ProcessSensorMetadata(data_request->mutable_sensor_metadata());
+  for (cartographer::mapping::proto::Submap &mutable_submap :
```

Comment:
Can you add a comment here, since this is easy to miss: Something like that the SubmapIds also contain trajectory IDs and those need to be translated as well to be correctly attributed to the trajectory in the cloud?

Reply:
Done